pkg/token/service.go

```diff
@@ -423,72 +438,105 @@ func (d *tokenService) updateToken(key CacheKey, tt mode) error {
+	for _, k := range d.rtTargetDomainRoles {
+		wg.Add(1)
+		go func(k CacheKey) {
+			defer wg.Done()
+			domain, role := k.Domain, k.Role
+			token := d.roleTokenCache.Load(k)
+			outPath := filepath.Join(d.tokenDir, domain+":role."+role+".roletoken")
+			err := d.writeFileWithRetry(ctx, maxElapsedTime, token, outPath, mROLE_TOKEN)
+			if err != nil {
+				echan <- err
+				rtErrorCount.Add(1)
+			}
+		}(k)
 	}
-	err = d.roleTokenCache.Range(func(k CacheKey, t Token) error {
-		domain := t.Domain()
-		role := t.Role()
-		rt := t.Raw()
-		if rt == "" {
-			// skip placeholder token added during daemon creation
-			return nil
-		}
 
-		log.Infof("[New Role Token] Domain: %s, Role: %s", domain, role)
-		outPath := filepath.Join(d.tokenDir, domain+":role."+role+".roletoken")
-		log.Debugf("Saving Role Token[%d bytes] at %s", len(rt), outPath)
-		if err := w.AddBytes(outPath, 0644, []byte(rt)); err != nil {
-			return fmt.Errorf("unable to save role token: %w", err)
-		}
+	// wait for ALL token updates to complete
+	wg.Wait()
+	log.Infof("Saved tokens to files. accesstoken:success[%d],error[%d]; roletoken:success[%d],error[%d]", int64(len(d.atTargetDomainRoles))-atErrorCount.Load(), atErrorCount.Load(), int64(len(d.rtTargetDomainRoles))-rtErrorCount.Load(), rtErrorCount.Load())
+
+	// collect errors
+	close(echan)
+	errs := make([]error, 0, len(d.atTargetDomainRoles)+len(d.rtTargetDomainRoles))
+	for err := range echan {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+func (d *tokenService) writeFileWithRetry(ctx context.Context, maxElapsedTime time.Duration, token Token, outPath string, tt mode) error {
+	// backoff config with first retry delay of 5s, and then 10s, 20s, ...
+	b := backoff.NewExponentialBackOff()
+	b.InitialInterval = 5 * time.Second
+	b.Multiplier = 2
+	b.MaxElapsedTime = maxElapsedTime
+
+	operation := func() error {
+		return d.writeFile(token, outPath, tt)
+	}
+	notifyOnErr := func(err error, backoffDelay time.Duration) {
+		log.Errorf("Failed to output the token to a file: %s. Retrying in %s", err.Error(), backoffDelay)
+	}
+	return backoff.RetryNotify(operation, backoff.WithContext(b, ctx), notifyOnErr)
+}
+
+func (d *tokenService) writeFile(token Token, outPath string, tt mode) error {
+	w := util.NewWriter()
+	tokenType := ""
+	switch tt {
+	case mACCESS_TOKEN:
+		tokenType = "Access"
+	case mROLE_TOKEN:
+		tokenType = "Role"
+	default:
+		return fmt.Errorf("invalid token type: %d", tt)
+	}
+
+	domain := token.Domain()
+	role := token.Role()
+	rawToken := token.Raw()
+	if rawToken == "" {
+		// skip placeholder token added during daemon creation
 		return nil
-	})
-	if err != nil {
-		return err
 	}
 
+	log.Infof("[New %s Token] Domain: %s, Role: %s", tokenType, domain, role)
+	log.Debugf("Saving %s Token[%d bytes] at %s", tokenType, len(rawToken), outPath)
+	if err := w.AddBytes(outPath, 0644, []byte(rawToken)); err != nil {
+		return fmt.Errorf("unable to save %s Token: %w", tokenType, err)
+	}
 	return w.Save()
 }
```

Comment:
Again, try to avoid refactoring in an implementation PR.

You can just write a simple TODO saying,
// TODO: Refactor writing file function as it is the same for RT/AT

And create a new refactor PR based on it. (or an issue)